src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs

```diff
@@ -20,6 +20,31 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
             return thisType.CanCastToInternal(otherType, null);
         }
 
+        public static bool IsCompatibleWith(this TypeDesc thisType, TypeDesc otherType)
+        {
+            // Structs can be cast to the interfaces they implement, but they are not compatible according to ECMA I.8.7.1
+            bool isCastFromValueTypeToReferenceType = otherType.IsValueType && !thisType.IsValueType;
+            if (isCastFromValueTypeToReferenceType)
+            {
+                return false;
+            }
+
+            // Nullable<T> can be cast to T, but this is not compatible according to ECMA I.8.7.1
+            bool isCastFromNullableOfTtoT = thisType.IsNullable && otherType.IsEquivalentTo(thisType.Instantiation[0]);
+            if (isCastFromNullableOfTtoT)
+            {
+                return false;
+            }
+
+            return otherType.CanCastTo(thisType);
+        }
+
+        private static bool IsEquivalentTo(this TypeDesc thisType, TypeDesc otherType)
+        {
+            // TODO: Once type equivalence is implemented, this implementation needs to be enhanced to honor it.
```

Comment:
Is type equivalence used for anything outside COM?

We don't intend to model COM in the managed type system so if this is exclusive to COM, the use of this can be replaced with `==` (we can safely use refence equality in the type system; no need to call `Equals`).

Reply:
I think that it is used just by COM, @davidwrighton can you please confirm that?